docs: add file uploads (#4626)
apollographql/router: docs/source/executing-operations/file-uploads.mdx

## About file uploads using multipart requests

The [GraphQL Multipart Request Spec](https://github.com/jaydenseric/graphql-multipart-request-spec) enables you to send files using arguments in a GraphQL mutation.

{/* To-do: specifics + examples */}

## File upload configuration and usage

To enable file uploads in clients, you must both [configure support in the Apollo Router](#configure-file-upload-support-in-the-router) and [ensure client usage conforms to requirements](#client-usage-requirements).

### Configure file upload support in the router

By default, receiving client file uploads is _not_ enabled in the Apollo Router.
To enable file upload support, set the following fields in your `router.yaml` configuration file:

```yaml title="router.yaml" showLineNumbers="false"
preview_file_uploads:
  enabled: true
  protocols:
    multipart:
      enabled: true
      mode: streaming
      limits:
        max_file_size: 5mb # default 1mb
        max_files: 2 # default 4
```

#### Mode

The only supported `mode` is `streaming`. This mode means that the router doesn't retain uploaded files in memory during a request.

> If a request cannot be fulfilled in a streaming fashion, the router returns the [`UPLOADS_OPERATION_CANNOT_STREAM`](#uploads_operation_cannot_stream) error.

Reviewer: Let's add some brief guidance on what will definitely work for people. Something like:

To ensure that your operation is streamable, make sure to avoid nesting for file uploads.

For example (fixme):

```
Mutation {
  upload_file()
  upload_file()
  upload_file()
}
```

Author: Good call. Beyond nesting files, are we are of anything that would preclude an operation from being streamable? There's already a size limit and I'm not sure it's necessary to say anything abut network connections, so maybe not?

Author: Also, could you please check the now included example makes sense? TIA!

Reviewer: It's more like the example you have is OK but further nesting is not recommended.
```
mutation($f1: File, $f2: File, $f3: File) {
    upload(data: $f1) {
        __typename
    }
    upload(data: $f2) {
        __typename
    }

    somethingComplex{
        upload(data: $f3) {
            __typename
        }
    }
}

```
`f1` and `f2` are OK. They are top level in the mutation.
`f3` is nested in something else, and is more risky.

Author: I see. Thank you!

---
title: File uploads
description: Receive file uploads with the Apollo Router

> minVersion: X.x  

Author: To-do: Add router version

Author: Is 1.41.0 the correct version?

```suggestion
minVersion: 1.41.0  
```

Reviewer 2: I believe that will be correct

preview_file_uploads:
  enabled: true
  protocols:
    multipart:
      enabled: true
      mode: streaming
      limits:
        max_file_size: 5mb # default 1mb
        max_files: 2 # default 4
```

#### Mode

The only supported `mode` is `streaming`.
That means the router doesn't retain uploaded files in memory during a request.
Streaming file uploads can be more memory-efficient, especially for large files, since it avoids loading the entire file into memory.

To ensure your operation is streamable, avoid nesting in file uploads.
For example, the following operation would not be streamable:

```graphql title="❌" disableCopy=true showLineNumbers=false
mutation UploadNestedFiles($files: [Upload!]!) {
  uploadNestedFiles(files: $files) {
    success
  }
}

```

If a request cannot be fulfilled in a streaming fashion, the router returns the [`UPLOADS_OPERATION_CANNOT_STREAM`](#uploads_operation_cannot_stream) error.

#### Limits

The router includes default limits for file uploads to prevent denial-of-service attacks.
You can configure both the maximum file size and number of files to accept.
If a request exceeds limits, the router rejects them.

#### Configuration reference

> 

Reviewer 3: Mention `preview_file_uploads` here or in the table header as the root option

<table class="field-table">
<thead>
<tr>
<th>Attribute/ <br/> Default value</th>
<th>Description</th>
<th>Valid Values</th>
</tr>
</thead>
<tbody>
<tr>
<td>

##### `enabled`

Default: `false`

</td>
<td>Flag to enable reception of client file uploads</td>
<td>boolean</td>

</tr>
<tr>
<td>

##### `protocols.multipart.enabled`
Default: `false`

</td>
<td>Flag to enable reception of multipart file uploads</td>
<td>boolean</td>
</tr>
<tr>
<td>

##### `protocols.multipart.mode`

> Default: None

Reviewer 3: Should `streaming` be the default, at least until there's another mode?

Author: Good question for the team—maybe it's even already the case?

@BrynCooke or @IvanGoncharov would you be able to confirm the default for `mode`? If there is no default, wouldn't it be safer to make it `streaming` as that's the required value?

Reviewer: I think `streaming` should be mandatory in case we decide to add a new mode later and have it be default.

</td>
<td>Supported file upload mode</td>
<td> 

`streaming`
(required value)

</td>
</tr>
<tr>
<td>

##### `protocols.multipart.limits.max_file_size`

> Default: `1mb`

Reviewer 3: @BrynCooke - should `mb` be `MB` (or `MiB` if more accurate)? Should've asked in other PRs about whether we standardize on our notation for units of size

Author: I also commented this in the draft docs, but I'm curious about the units we accept.

Reviewer: We use https://crates.io/crates/bytesize, which supports a variety of different units.

</td>

> <td>The maximum file size to accept</td>

Reviewer 3: Reiterate consequence of exceeding limit

```suggestion
<td>The maximum file size to accept.
If this limit is exceeded, the router rejects the entire request.
</td>
```

<td>

Accepts values in a human-readable format; for example, `5kb` and `99mb` are acceptable values.

</td>
</tr>
<tr>
<td>

##### `protocols.multipart.limits.max_files`
Default: `4`

</td>

> <td>The maximum number of files to accept</td>

Reviewer 3: ```suggestion
<td>The maximum number of files to accept.
If this limit is exceeded, the router rejects the entire request.
</td>
```

## Metrics for file uploads

Metrics in the Apollo Router for file uploads:

<table class="field-table metrics">
  <thead>
    <tr>
      <th>Name</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr>
<td>

##### `apollo.router.operations.file_uploads`

</td>
<td>

Counter for the number of file uploads.

</td>
</tr>

<tr>
<td>

##### `apollo.router.operations.file_uploads.file_size`

</td>
<td>

Histogram for the size of uploaded files.

</td>
</tr>

<tr>
<td>

##### `apollo.router.operations.file_uploads.files`

</td>
<td>

Histogram for the number of uploaded files.

</td>
</tr>

</tbody>
</table>

## Error codes for file uploads

A file upload request may receive the following error responses:

<table class="field-table metrics">
<tr>
<th>Error Code</th>
<th>Description</th>
</tr>
<tr>
<td>

##### `UPLOADS_LIMITS_MAX_FILES_EXCEEDED`

</td>
<td>The number of files in the request exceeded the configured limit</td>
</tr>
<tr>
<td>

##### `UPLOADS_LIMITS_MAX_FILE_SIZE_EXCEEDED`

</td>

> <td>A file exceeded the maximum configured file size.</td>

Reviewer 3: ```suggestion
<td>A file exceeded the maximum configured file size</td>
```

```yaml title="router.yaml" showLineNumbers="false"
preview_file_uploads:
  enabled: true
  protocols:
    multipart:
      enabled: true
      mode: streaming
      limits:
        max_file_size: 5mb # default 1mb
        max_files: 2 # default 4
```

#### Mode

The only supported `mode` is `streaming`.
That means the router doesn't retain uploaded files in memory during a request.
Streaming file uploads can be more memory-efficient, especially for large files, since it avoids loading the entire file into memory.

To ensure your operation is streamable, avoid nesting file uploads.
For example, the following `nestedUpload` operation attempting to upload `$file3` would not be streamable:

```graphql title="❌" disableCopy=true showLineNumbers=false
mutation uploadFiles($file1: Upload, $file3: Upload, $file3: Upload ) {
  upload(data: $file1) { 
    __typename
  }
  upload(data: $file2) {
    __typename
  }
  
  nestedUpload {
    upload(data: $file3) {
      __typename
    }
  }
}
```

If a request cannot be fulfilled in a streaming fashion, the router returns the [`UPLOADS_OPERATION_CANNOT_STREAM`](#uploads_operation_cannot_stream) error.

> 

Author: ```suggestion

<Caution>

To avoid streaming issues, you cannot use file uploads with [AWS Signature Version 4 (SigV4) authentication](../configuration/authn-subgraph/).

</Caution>

```
@BrynCooke , could you confirm this is an accurate representation of the sigv4 limitation?

Reviewer: Yes that is correct.

<tr>
<td>

##### `apollo.router.operations.file_uploads.file_size`

</td>
<td>

Histogram for the size of uploaded files.

</td>
</tr>

<tr>
<td>

##### `apollo.router.operations.file_uploads.files`

</td>
<td>

Histogram for the number of uploaded files.

</td>
</tr>

</tbody>
</table>

## Error codes for file uploads

A file upload request may receive the following error responses:

<table class="field-table metrics">
<tr>
<th>Error Code</th>
<th>Description</th>
</tr>
<tr>
<td>

##### `UPLOADS_LIMITS_MAX_FILES_EXCEEDED`

</td>
<td>The number of files in the request exceeded the configured limit</td>
</tr>
<tr>
<td>

##### `UPLOADS_LIMITS_MAX_FILE_SIZE_EXCEEDED`

</td>
<td>A file exceeded the maximum configured file size</td>
</tr>
<tr>
<td>

##### `UPLOADS_FILE_MISSING`

</td>
<td>The operation specified a file that was missing from the request</td>
</tr>
<tr>
<td>

##### `UPLOADS_OPERATION_CANNOT_STREAM`

</td>
<td>The request was invalid as it couldn't be streamed to the client</td>
</tr>
</table>

## Known limitations

While in private preview, Apollo recommends using file uploads only in development and testing environments, not in production.

### Unsupported query modes

> When file uploads are enabled in the router, any operations using the [`@defer`](/graphos/operations/defer/) are unsupported.

Author: ```suggestion
The router rejects operations that use file upload variables on or inside fields using [`@defer`](/graphos/operations/defer/).

```graphql title="❌ Unsupported usage" disableCopy=true showLineNumbers=false
mutation CreatePost($title: String!, $body: String! $image: Upload!) {
  createPost(title: $title, body: $body, image: $image) {
    id
    title
    body
    image @defer
  }
}
```

```graphql title="✅ Supported usage" disableCopy=true showLineNumbers=false
mutation CreatePost($title: String!, $body: String! $image: Upload!) {
  createPost(title: $title, body: $body, image: $image) {
    id
    title
    body @defer
    image
  }
}
```
```
